#164198143 Users can reset password

fejiroofficial opened the pull request:
**What does this PR do?**
 This is PR for:
- Building a feature that allows users to reset a forgotten password

**Description of Task to be completed?**
- Write unit test(s)
- Write reducers and actions to be executed 
- Write components and validators for feature

**How should this be manually tested**
- Clone this repo and cd into it
- run `npm install` to install dependencies.
- test using jest with `npm test`. 

**What are the relevant pivotal tracker stories**
[#164198143](https://www.pivotaltracker.com/n/projects/2313263/stories/164198143 )

**Screenshots**
![ezgif com-video-to-gif (2)](https://user-images.githubusercontent.com/38490848/56182799-b3dca780-600b-11e9-88f5-941383b77a61.gif)

<img width="1422" alt="Screenshot 2019-04-16 at 6 14 11 AM" src="https://user-images.githubusercontent.com/38490848/56183594-09ff1a00-600f-11e9-900f-d60df58dd043.png">

aanchirinah commented on src/actions/actionTypes.js:
```diff
@@ -0,0 +1,7 @@
+export const ADD_DATA = 'ADD_DATA'; // dummy constant
+
+export const PASSWORD_RESET_REQUEST = 'PASSWORD_RESET_REQUEST';
```
#### Nitpick
- Use an actionType generator.

fejiroofficial replied:
okay, I will work on this

fejiroofficial replied:
This has been taken care of now

aanchirinah commented on src/actions/resetPassword.js:
```diff
@@ -0,0 +1,44 @@
+/* eslint-disable react/require-extension */
+import axios from 'axios';
+import {
+  PASSWORD_RESET_REQUEST,
```
#### Nitpick
- Why not this?
```suggestion
  PASSWORD_RESET_REQUEST_LOADING,
```

fejiroofficial replied:
I felt it was quite long, but since it is more descriptive, I will return to this.

aanchirinah commented on src/actions/resetPassword.js:
```diff
@@ -0,0 +1,44 @@
+        type: PASSWORD_RESET_REQUEST_SUCCESS,
+        isLoading: false
+      });
+      return data;
```
#### Nitpick
- Why are you returning `data`?

fejiroofficial replied:
This has been fixed and sent to the store in a new action dispatch

aanchirinah commented on src/components/ResetPassword/index.jsx:
```diff
@@ -0,0 +1,95 @@
+  const input = useRef(null);
+  
+  const updateInput = event => {
+    setUserEmail({ ...userEmail, [event.target.name]: event.target.value });
```
#### Nitpick
- Why are you using spread operator here?

fejiroofficial replied:
The state used here is an object and I think it is a good practice to not mutate it directly.

aanchirinah commented on src/components/ResetPassword/index.jsx:
```diff
@@ -0,0 +1,95 @@
+  };
+
+  return (
+    <Modal clearMessage={() => setMessage(null)} title="Recover Password!">
```
#### Nitpick
- What happened to the pattern of having views and container component?

fejiroofficial replied:
There are two components involved, the modal is a view while this one is a container

aanchirinah commented on src/components/ResetPassword/index.jsx:
```diff
@@ -0,0 +1,95 @@
+
+  const handleSubmit = () => {
+    setMessage(null);
+    props.passwordResetRequest(userEmail.email).then(res => {
```
#### Nitpick
- This is anti-pattern with regards to flux architecture. Subscribe to whatever it is you need in the `redux store` and use that to handle this.

fejiroofficial replied:
okay! this has been taken care of now.

aanchirinah commented on src/reducers/resetPasswordReducer.js:
```diff
@@ -0,0 +1,31 @@
+import {
+  PASSWORD_RESET_REQUEST_FAILURE,
```
#### Nitpick
- Use selectors

fejiroofficial replied:
I will work on this now.

aanchirinah commented on src/reducers/resetPasswordReducer.js:
```diff
@@ -0,0 +1,31 @@
+export default function resetPasswordReducer(state = initialState, action = {}){
+  switch (action.type) {
+  case PASSWORD_RESET_REQUEST:
+    return { ...state, isLoading: action.isLoading};
```
#### Comment
- Why `spread` operator instead of `Object.assign()`

fejiroofficial replied:
okay, this is a matter of preference.

aanchirinah commented:
@fejiroofficial  What is the status of this?

aanchirinah commented:
@fejiroofficial I don't see it being done from the component @thislekan created,  any reason why?

fejiroofficial commented:
@aanchirinah That was the old one, it has been updated.

valenstical commented:
@fejiroofficial, you have already implemented your feature from the component @thislekan, created. Simply update the GIF image in this PR to show you clicking the 'forgot password' button inside the login form.

fejiroofficial commented:
@valenstical It has been updated. I added a screenshot of the email received too and fixed the duplicated code that code climate is complaining about.

valenstical commented on src/reducers/resetPasswordReducer.js:
```diff
@@ -0,0 +1,47 @@
+}
+
+
+// export default function resetPasswordReducer(state = initialState, action = {}) {
```
Moving forward, do you think you can remove unused codes from the codebase?
